R/CompDbSource.R

```diff
@@ -221,196 +293,258 @@ setClass("MetFragParam",
 setMethod(
 	"matchSpectra", signature(query = "Spectra", target = "MetFragSource",
                               param = "MetFragParam"),
-	function(query, target, param, BPPARAM = BiocParallel::SerialParam()) {
-        out <- list()
-		names(out) <- spectraData(query)$peak_id
-        mtch <- data.frame()	
+	function(query, target, param, debug = FALSE) {
+        validAdducts(query, param)
+        if (debug) handle <- verbose()
+        else handle <- NULL
+        out <- vector("list", length(query))
+        restQuery <- .param_to_list(param)
+        pb <- progress_bar$new(format = paste0("[:bar] :current/:",
+                                               "total (:percent) in ",
+                                               ":elapsed"),
+                               total = length(query), clear = FALSE)
+        pb$tick(0)
 		for (i in seq_along(query)) {
-			## get peaks data from query[i] as list/character string for MetFrag (assuming single peaks table per spectra - is it always valid?!)	
-			peaks <- peaksData(query[i])[[1]]
-			## We allow either 'minFragCount' number of fragment peaks in spectra (under important uncontrolled auumption of proper data pretreatment), OR 
-			##	a 
-			if (nrow(peaks) < mfParam@minFragCount & nrow(peaks) > 0) {
-				## In cases of very few fragment peaks, the MetFrag server is 'timing-out' due (probably) to the large search space -
-				##	thus, I think the it's better to send the query with an empty fragment spectra and at least get the basic initial search matrix.
-				peaks <- peaks[-c(1:nrow(peaks)),]
-				## ...the other option is to ignor these calls - leaving the input as-is and returning empty results:
-				# out[[i]] = list(
-					# mfResults = NULL,
-					# mfServerStatus = "NA"
-				# )
-				# next()				
-			}
-			
-			peakListString <- paste(apply(peaks, 1, function (x){ paste(x[1],x[2],sep = '_') }),collapse = ';')
-
-			mz <- spectraData(query[i])$precursorMz
-
-			chkAdducts <- validAdducts(query,param)	
-			
-			if (!all(chkAdducts)) { 
-				stop("Some spectra precursor adducts are not legal MetFrag adduct types: Use function 'validAdducts(MetFragParam())' to print a list of valid adduct names. 'matchSpectra' stopping.") 
-			}
-							
-			adduct <- spectraData(query)$precursorAdduct[i]
-			## All REST calls will be done inside this loop here:
-			out[[i]] <- lapply(1:length(adduct),function (i) {
-				neutralPrecursorMass <- MetaboCoreUtils::mz2mass(mz, adduct[i])
-				restQuery <- sapply(slotNames(param)[-which(slotNames(param) %in% c("validAdducts","THRESHFUN"))], function(x) { 
-					slot(param,x)
-				})
-				restQuery <- c(
-					restQuery,	
-					"neutralprecursormass" = as.character(neutralPrecursorMass),
-					"precursoriontype" = as.character(adduct),
-					"peakliststring" = peakListString
+            qi <- query[i]
+			peaks <- peaksData(qi)[[1L]]
+			if (nrow(peaks) < param@minFragCount) {
+                out[[i]] <- data.frame(
```

Comment:
Note: with this version we report ALWAYS something for a query:

- `server_status = NA`: one row in the result `data.frame` with all `NA`s.
- `server_status = "TIMEOUT"`: query was sent but a timeout was reached - or some other error from the server. one row in the result `data.frame` with all `NA`s.
- `server_status = "SUCCESS"`: query successfull, x rows in the result `data.frame`.